packages/make-hardener/src/main.js

```diff
@@ -33,113 +59,216 @@ const { ownKeys } = Reflect;
+      Object,
+      Object.getOwnPropertyDescriptors(patchObjectMethods),
+    );
+
+    defineProperties(
+      Reflect,
+      Object.getOwnPropertyDescriptors(patchReflectMethods),
+    );
+  }
+
+  /**
+   * if ENABLE_ONLY_NON_EXTENSIBLE then this will be the
+   * patched one. Otherwise it will be the original.
+   */
+  const { freeze: patchedObjectFreeze } = Object;
 
-      function dequeue() {
-        // New values added before forEach() has finished will be visited.
-        toFreeze.forEach(freezeAndTraverse); // todo curried forEach
+  /**
+   * Currently not exported. But we should consider exporting it
+   * as a SES peer of isFrozen, etc. Of course, we'd need to remove
+   * the `path` parameter first.
+   *
+   * @param {any} val
+   * @param {string=} path
+   * @returns {boolean}
+   */
+  const isHardened = (val, path) => {
+    if (val !== Object(val)) {
+      return true;
+    }
+    let answer1;
+    let answer2;
+    if (ENABLE_OLD_HARDENED) {
+      answer1 = oldHardened.has(val);
+    }
+    if (ENABLE_ONLY_NON_EXTENSIBLE) {
+      // We check isFrozen, but we accumulate objects when they become
+      // non-extensible, because once non-extensible they might become frozen,
+      // and we can more feasibly catch all operations that might make them
+      // non-extensible.
+      answer2 = isFrozen(val) && !onlyNonExtensible.has(val);
+    }
+    if (ENABLE_OLD_HARDENED && ENABLE_ONLY_NON_EXTENSIBLE) {
+      if (answer1 !== answer2) {
+        // Not necessarily an error. Might be one of the three safe
+        // exceptions explained in the README.md. We should test these
+        // outliers, because if it is anything else, we may have a
+        // serious bug.
+        console.log(`HARDEN disagreement: ${answer1} vs ${answer2} at ${path}`);
       }
+    }
 
-      function commit() {
-        // todo curried forEach
-        // we capture the real WeakSet.prototype.add above, in case someone
-        // changes it. The two-argument form of forEach passes the second
-        // argument as the 'this' binding, so we add to the correct set.
-        toFreeze.forEach(hardened.add, hardened);
+    return !!(answer1 || answer2);
+  };
+
+  const moreObjectsHardened = toHarden => {
+    if (ENABLE_OLD_HARDENED && ENABLE_ONLY_NON_EXTENSIBLE) {
+      // TODO What consistency check should we put here?
+    }
+    if (ENABLE_OLD_HARDENED) {
+      toHarden.forEach(weakSetAdd, oldHardened);
+    }
+    if (ENABLE_ONLY_NON_EXTENSIBLE) {
+      toHarden.forEach(weakSetDelete, onlyNonExtensible);
+    }
+  };
+
+  /**
+   * @template T
+   * @param {T} root
+   * @returns {T}
+   */
+  const harden = root => {
+    const toHarden = new Set();
+    const paths = new WeakMap();
+
+    // If val is something we should be freezing but aren't yet,
+    // add it to toFreeze.
+    /**
+     * @param {any} val
+     * @param {string} [path]
+     */
+    function enqueue(val, path = undefined) {
+      if (Object(val) !== val) {
```

Comment:
can we do something here that doesn't allocate?  Failing that, I notice this same test in `isHardened`. We should consolidate that.

Reply:
I believe the 0-alloc equivalent is `typeof val === 'object' && val !== null`. I distinctly remember a TC-39 where markm showed `Object(val) === val` and Doug Crockford said something like, “This code sample must not leave this room,” which was at least on-brand.

Reply:
Just noting here that if records and tuples are both
   * primitives
   * can contain mutable objects

then there is no way to write such code now that will remain correct over the introduction of records and tuples. Attn @littledan